components/engine/src/lib.rs

```diff
@@ -0,0 +1,408 @@
+            errorpb
+                .mut_key_not_in_region()
+                .set_start_key(start_key.to_vec());
+            errorpb
+                .mut_key_not_in_region()
+                .set_end_key(end_key.to_vec());
+        }
+
+        errorpb
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Engines {
+    pub kv: Arc<DB>,
+    pub raft: Arc<DB>,
+}
+
+impl Engines {
+    pub fn new(kv_engine: Arc<DB>, raft_engine: Arc<DB>) -> Engines {
+        Engines {
+            kv: kv_engine,
+            raft: raft_engine,
+        }
+    }
+
+    pub fn write_kv(&self, wb: WriteBatch) -> Result<()> {
+        self.kv.write(wb).map_err(Error::RocksDb)
+    }
+
+    pub fn write_kv_opt(&self, wb: WriteBatch, opts: &WriteOptions) -> Result<()> {
+        self.kv.write_opt(wb, opts).map_err(Error::RocksDb)
+    }
+
+    pub fn sync_kv(&self) -> Result<()> {
+        self.kv.sync_wal().map_err(Error::RocksDb)
+    }
+
+    pub fn write_raft(&self, wb: WriteBatch) -> Result<()> {
+        self.raft.write(wb).map_err(Error::RocksDb)
+    }
+
+    pub fn write_raft_opt(&self, wb: WriteBatch, opts: &WriteOptions) -> Result<()> {
+        self.raft.write_opt(wb, opts).map_err(Error::RocksDb)
+    }
+
+    pub fn sync_raft(&self) -> Result<()> {
+        self.raft.sync_wal().map_err(Error::RocksDb)
+    }
+}
+
+// TODO: refactor this trait into rocksdb trait.
+pub trait Peekable {
+    fn get_value(&self, key: &[u8]) -> Result<Option<DBVector>>;
+    fn get_value_cf(&self, cf: &str, key: &[u8]) -> Result<Option<DBVector>>;
+
+    fn get_msg<M: protobuf::Message>(&self, key: &[u8]) -> Result<Option<M>> {
+        let value = self.get_value(key)?;
+
+        if value.is_none() {
+            return Ok(None);
+        }
+
+        let mut m = M::new();
+        m.merge_from_bytes(&value.unwrap())?;
+        Ok(Some(m))
+    }
+
+    fn get_msg_cf<M: protobuf::Message>(&self, cf: &str, key: &[u8]) -> Result<Option<M>> {
+        let value = self.get_value_cf(cf, key)?;
+
+        if value.is_none() {
+            return Ok(None);
+        }
+
+        let mut m = M::new();
+        m.merge_from_bytes(&value.unwrap())?;
+        Ok(Some(m))
+    }
+
+    fn get_u64(&self, key: &[u8]) -> Result<Option<u64>> {
+        let value = self.get_value(key)?;
+
+        if value.is_none() {
+            return Ok(None);
+        }
+
+        let value = value.unwrap();
+        if value.len() != 8 {
+            return Err(Error::Other(
+                format!("need 8 bytes, but only got {}", value.len()).into(),
+            ));
+        }
+
+        let n = BigEndian::read_u64(&value);
+        Ok(Some(n))
+    }
+
+    fn get_i64(&self, key: &[u8]) -> Result<Option<i64>> {
+        let r = self.get_u64(key)?;
+        match r {
+            None => Ok(None),
+            Some(n) => Ok(Some(n as i64)),
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+enum SeekMode {
+    TotalOrder,
+    Prefix,
+}
+
+pub struct IterOption {
```

Comment:
How about just keep it as `pub struct IterOption(ReadOption)`? So that `*_bound`s doesn't have to be stored twice.

Reply:
Better in another pr, this pr just split modules.